keras_nlp/models/whisper/whisper_backbone.py

```diff
@@ -0,0 +1,287 @@
+
+    This class implements a Transformer-based encoder-decoder model as
+    described in
+    ["Robust Speech Recognition via Large-Scale Weak Supervision"](https://arxiv.org/abs/2212.04356).
+    It includes the embedding lookups and transformer layers, but not the head
+    for predicting the next token.
+
+    The default constructor gives a fully customizable, randomly initialized Whisper
+    model with any number of layers, heads, and embedding dimensions. To load
+    preset architectures and weights, use the `from_preset` constructor.
+
+    Disclaimer: Pre-trained models are provided on an "as is" basis, without
+    warranties or conditions of any kind. The underlying model is provided by a
+    third party and subject to a separate license, available
+    [here](https://github.com/openai/whisper).
+
+    Args:
+        vocabulary_size: int. The size of the token vocabulary.
+        num_layers: int. The number of transformer encoder layers and
+            transformer decoder layers.
+        num_heads: int. The number of attention heads for each transformer.
+            The hidden size must be divisible by the number of attention heads.
+        hidden_dim: int. The size of the transformer encoding and pooler layers.
+        intermediate_dim: int. The output dimension of the first Dense layer in
+            a two-layer feedforward network for each transformer.
+        num_mels: int. The number of mel-frequency filters.
+        dropout: float. Dropout probability for the Transformer encoder.
+        max_source_sequence_length: int. The maximum sequence length that the
+            audio encoder can consume. Since the second convolutional layer in
+            the encoder reduces the sequence length by half (stride of 2), we
+            use `max_source_sequence_length // 2` as the sequence length for the
+            positional embedding layer.
+        max_target_sequence_length: int. The maximum sequence length that the
+            text decoder can consume.
+
+    Examples:
+    ```python
+    input_data = {
+        "encoder_token_ids": tf.ones(shape=(1, 12, 80), dtype=tf.int64),
+        "decoder_token_ids": tf.ones(shape=(1, 12), dtype=tf.int64),
+        "decoder_padding_mask": tf.constant(
+            [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0], shape=(1, 12)
+        ),
+    }
+
+    # Randomly initialized Whisper encoder-decoder model with a custom config
+    model = keras_nlp.models.WhisperBackbone(
+        vocabulary_size=51864,
+        num_layers=6,
+        num_heads=8,
+        hidden_dim=512,
+        intermediate_dim=2048,
+        num_mels=80,
+        max_source_sequence_length=128,
+        max_target_sequence_length=64,
+    )
+    output = model(input_data)
+    ```
+    """
+
+    def __init__(
+        self,
+        vocabulary_size,
+        num_layers,
+        num_heads,
+        hidden_dim,
+        intermediate_dim,
+        num_mels=80,
+        dropout=0.0,
+        max_source_sequence_length=3000,
```

Comment:
Should we name these `max_encoder_sequence_length` and `max_decoder_sequence_length`? It seems weird to have the encoder/decoder terminology everywhere but here.